src/utils/wallet/solana.ts

```diff
@@ -81,122 +81,216 @@ export function fetchOptions(chain: Chain) {
 // This function signs and sends the transaction while constantly checking for confirmation
 // and resending the transaction if it hasn't been confirmed after the specified interval
 // See https://docs.triton.one/chains/solana/sending-txs for more information
-export async function signAndSendTransaction(
+export async function signAndSendTransactionWithRetry(
   request: SolanaUnsignedTransaction<Network>,
   wallet: Wallet | undefined,
   options?: ConfirmOptions,
-) {
+): Promise<string> {
   if (!wallet) throw new Error('Wallet not found');
   const rpc = config.rpcs[request.chain];
   if (!rpc) throw new Error(`${request.chain} RPC not found`);
 
   const commitment = options?.commitment ?? 'finalized';
   const connection = new Connection(rpc);
+
   const { blockhash, lastValidBlockHeight } =
     await connection.getLatestBlockhash(commitment);
 
-  const unsignedTx = await setPriorityFeeInstructions(
+  const { serializedTransaction, sendOptions } = await createSolanaTransaction(
+    request,
+    wallet,
     connection,
     blockhash,
     lastValidBlockHeight,
-    request,
+    commitment,
   );
 
-  let confirmTransactionPromise: Promise<
-    RpcResponseAndContext<SignatureResult>
-  > | null = null;
-  let confirmedTx: RpcResponseAndContext<SignatureResult> | null = null;
-  let txSendAttempts = 1;
-  let signature = '';
-  // TODO: VersionedTransaction is supported, but the interface needs to be updated
-  const tx = await wallet.signTransaction(unsignedTx as Transaction);
-  const serializedTx = tx.serialize();
-  const sendOptions = {
-    skipPreflight: true,
-    maxRetries: 0,
-    preFlightCommitment: commitment, // See PR and linked issue for why setting this matters: https://github.com/anza-xyz/agave/pull/483
-  };
-  signature = await connection.sendRawTransaction(serializedTx, sendOptions);
-  confirmTransactionPromise = connection.confirmTransaction(
+  const signature = await connection.sendRawTransaction(
+    serializedTransaction,
+    sendOptions,
+  );
+
+  const newSignature = await confirmTransactionWithRetry(
+    signature,
+    { serializedTransaction, sendOptions },
+    connection,
+    blockhash,
+    lastValidBlockHeight,
+    commitment,
+  );
+
+  if (!newSignature) throw new Error('Transaction failed');
+
+  return newSignature;
+}
+
+async function confirmTransactionWithRetry(
+  signature: string,
+  transaction: {
+    serializedTransaction: Uint8Array | Buffer | number[];
+    sendOptions?: SendOptions;
+  },
+  connection: Connection,
+  blockHash: string,
+  lastValidBlockHeight: number,
+  commitment?: Commitment,
+) {
+  let currentSignature = signature;
+  let confirmTransactionPromise = await connection.confirmTransaction(
```

Comment:
We should not use `await` here, since `confirmTransaction` blocks until the transaction is either confirmed or expired, and we won’t be resending the transaction in the meantime. Instead, we should call `confirmTransaction` *once* in this function and periodically check whether it has resolved or rejected *while also* resending the transaction. This improves the chances of the transaction being confirmed during periods of high congestion on Solana. Why are we changing this from the previous approach? I don’t think I understand what issue this PR is meant to address. See this article for why the previous approach is implemented: https://docs.triton.one/chains/solana/cascade/sending-txs

Reply:
We are resending the same transaction, it would not result in two independent transactions. The resent transactions all have the same signature as the original.